rsconnect/api.py

```diff
@@ -605,17 +637,31 @@ def check_server_capabilities(self, capability_functions):
                 raise RSConnectException(message)
         return self
 
+    def upload_lucid_bundle(self, prepare_deploy_result, bundle_size: int, contents):
+        upload_url = prepare_deploy_result.presigned_url
+        parsed_upload_url = urlparse(upload_url)
+        with S3Client("{}://{}".format(parsed_upload_url.scheme, parsed_upload_url.netloc), timeout=120) as s3_client:
+            upload_result = s3_client.upload(
+                "{}?{}".format(parsed_upload_url.path, parsed_upload_url.query),
+                prepare_deploy_result.presigned_checksum,
+                bundle_size,
+                contents,
+            )
+            S3Server(upload_url).handle_bad_response(upload_result)
+
     @cls_logged("Deploying bundle ...")
     def deploy_bundle(
         self,
         app_id: int = None,
+        project_application_id: str = None,
```

Comment:
There's sorta two separate workflows. 

1) External to cloud you can write code locally and then publish you code to cloud.
2) Internal to cloud you can have a project - and publishing the code in that project to cloud.

In cloud we create a link in the 2nd case between the project that has the code - and the output that ends up being the published version of the project. This supported through an environment variable `LUCID_APPLICATION_ID` which is present in every project - and consulted when `rsconnect` publishes an output.

So that's kinda the use case around this. I'm not sure that exposing this to an end user on the command line has much utility. It doesn't make a lot of sense for a person publishing locally to establish a link between the output that they are publishing - and some other random project in our system which doesn't have the code that they published. We also don't expose the projects application ID anywhere visible to the user...

I might not support this on the command line at all. If `LUCID_APPLICATION_ID` happens to be present - include it in the request. Otherwise don't try to make a link between a project and an output.